drop glog from the unquote and parser packages

mvdan opened the pull request:
(see commit messages - please do not squash)

Fixes #70.

cc @stapelberg @kssilveira

mvdan commented:
This is an alternative to https://github.com/protocolbuffers/txtpbfmt/pull/75. Note that, unlike the other patch, this avoids any breaking changes where possible.

There is one minor breaking change. Any `parser` users which did want glog should now pass it along explicitly, just like I'm doing in `cmd/txtpbfmt`. It can't be kept there by default, because we want to drop the hard dependency in the first place.

kssilveira commented:
I think this will work, I will comment on the diff.

kssilveira commented on parser/parser.go:
```diff
@@ -163,7 +173,9 @@ func Format(in []byte) ([]byte, error) {
 	addMetaCommentsToConfig(in, &c)
 	if c.Disable {
-		log.Infoln("Ignored file with 'disable' comment.")
+		if c.Logger != nil {
```
Can we avoid checking for nil in every use, and instead make every public function set logger to an internal noopLogger (that doesn't do anything) if logger is nil?

mvdan replied:
There's no good place to have such a read-write field to set to a noopLogger. The parser type isn't enough, because we need to log from some funcs which don't hang off the parser.

I assume you didn't like the repetition, so I set up two methods on the config type to deduplicate the nil checks. I don't think the nil checks harm performance, because an interface call always has to check whether the interface is nil or not.

kssilveira commented on parser/parser.go:
```diff
@@ -322,9 +334,9 @@ func parseWithMetaCommentConfig(in []byte, c Config) ([]*ast.Node, error) {
 	if err != nil {
 		return nil, err
 	}
-	if p.log {
```
Some logs, like this one, have V(2) by default, but others don't, so we need to somehow make the logger differentiate between "default verbosity" and "extra verbosity", such as having a boolean EnableExtraVerbosity.

mvdan replied:
I didn't even notice that there were actually three log levels, not just two. There's Errorf, Infof, _and_ the verbose Infof. Quite confusing to be honest, but I'm also not used to glog :) Naively, I would call the third a Debugf in the interface.

I've taken you up on the extra boolean for the sake of unblocking the PR.

kssilveira replied:
Sorry, I was wrong, there is only one verbosity, the reason for the separate `if` statement is the performance:

https://pkg.go.dev/github.com/golang/glog#V

`The second form is shorter but the first is cheaper if logging is off because it does not evaluate its arguments.`

Please delete `ExtraVerbosity` and instead use something like `if p.config.log()` implemented as `c.Logger.IsVerbose()` (c *Config) implemented as `return bool(l.Verbose)` (l parserLog).

I've checked that this compiles using https://go.dev/play/p/e8pZ220VUGK.

kssilveira reviewed (changes requested):
I wanted to approve only one commit, not the whole pull request, sorry.

kssilveira commented:
I think this is what we need. The difference to glog is the bool type,
which I think we can't do for an interface.

On Thu, Jan 5, 2023, 9:31 AM Daniel Martí ***@***.***> wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In parser/parser.go
> <https://github.com/protocolbuffers/txtpbfmt/pull/77#discussion_r1062219168>
> :
>
> > @@ -322,9 +334,9 @@ func parseWithMetaCommentConfig(in []byte, c Config) ([]*ast.Node, error) {
>  	if err != nil {
>  		return nil, err
>  	}
> -	if p.log {
>
> So we're ending up with this logging interface?
>
> // Logger is a small glog-like interface.
> type Logger interface {
> 	// Infof is used for informative messages, for testing or debugging.
> 	Infof(format string, args ...interface{})
>
> 	// Errorf is used for non-fatal errors, such as invalid meta comments.
> 	Errorf(format string, args ...interface{})
>
> 	// IsVerbose reports whether Infof logs are being printed out.
> 	IsVerbose() bool
> }
>
> Not opposed per se, though the interface stops being "small" or
> "glog-like" at that point :)
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/protocolbuffers/txtpbfmt/pull/77#discussion_r1062219168>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AACDCIN2YEEHXZRVFZVIQ53WQ2BGXANCNFSM6AAAAAASI7A4C4>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

mvdan commented:
Should be ready now. I went for `InfoLevel() bool` rather than `IsVerbose() bool`, to clarify that it returns whether the info level at `Infof` is enabled or not. The notion of "verbose" seemed confusing to me, as it could mean other things such as how detailed the logs should be, rather than whether they are printed or not.

mvdan commented:
Since the Errorf PR was merged (thanks @kssilveira!), I've now rebased and updated this PR accordingly.

The `Errorf` method is gone, as it's no longer needed. And I've also removed `InfoLevel` - since we now only need a single logger method, we can check that the logger is non-nil to construct log messages.

mvdan commented:
We could also go further and have `Logger` be just a func signature type, but I think that could be a bit limiting in the future.

kssilveira commented:
I will accept this, thanks!

mvdan commented:
Excellent, let me know if I can help to get this merged :)

mvdan commented:
FYI @kssilveira @stapelberg, this was merged with a change from `interface{}` to `any` - which was never in my PR, as far as I can tell. That broke the build, as `any` requires `go 1.18` or later in your `go.mod`. Master's build appears to be broken for now.

kssilveira commented:
Yes, sorry, fixing.
